[GAP-4] Computation Manifest (#27)
golemfactory/golem-architecture: gaps/gap-4_comp_manifest/gap-4_comp_manifest.md

description:
author: mf (@mfranciszkiewicz)
status: Draft
type: Feature
---

## Abstract

This GAP introduces a manifest file specification and a set of GCRS (Golem
Computation Resource Standards) properties (herein after referred to as
"Computation Manifest"), which allow Requestors to declare the constraints of
computations they may execute on a Provider's machine.

## Motivation

The need for a Computation Manifest originates from the upcoming Internet
access feature, where each outbound request is routed via Provider's host
operating system. In order to prevent unwanted / illegal behavior on Provider's
behalf, she needs to remain in control of actions that Requestors can execute
on her machine.

With Computation Manifests, Requestors constrain themselves to
a certain set of allowed actions, to be negotiated with and approved by a
Provider. Requestors' actions will be verified against the Manifest during
computation.

## Specification

Computation Manifest can take one of the following forms:

- an embedded JSON / YAML object within an external configuration file,
- `golem.srv.comp.manifest` namespace properties, an extension to
  Golem Computation Resource Standards.

> 

Reviewer: I get the scenario with Computation Manifest included in the Demand - as it specifies the actions/commands the Requestor will want to run on the host.

What about the "external configuration file" scenario? What configuration file would that be? Where would it reside? Would it be included in some sort of a "Golem application metadata"? Is this a step towards a Golem application "gallery"?

Author: In most cases, the Computation Manifest will be embedded in the Payload Manifest, where CM is co-signed with image hashes. This becomes the basic, trusted way of using the image. 

Apart from that, we'd rarely expect for CM properties to appear in a Demand. If they do, they're meant to extend the original manifest. Providers may agree to these terms if their internal rule set allows, e.g., running `/bin/date` within images signed by authority `A`. Of course it's possible to specify the whole CM in the Demand and rely on Provider's internal rule set, because the CM is not signed by a trusted party.

The "external configuration file" is an exercise for developers extending the manifest and the tooling around it. Rephrasing, it should be possible to embed such manifest in some location specified in a Demand, so that the manifest could be extracted and verified

> The Computation Manifest is expected be included in a Computation Payload

Reviewer: ```suggestion
The Computation Manifest is expected to be included in a Computation Payload
```

Manifest ([GAP](https://github.com/golemfactory/golem-architecture/pull/28))
file, where its hash is co-signed with Payload hashes. If a Computation

> Manifest is included in the Demand (and thus, Agreement), it's role is to

Reviewer: ```suggestion
Manifest is included in the Demand (and thus, Agreement), its role is to
```

> derive the properties from and override the initially sourced Manifest

Reviewer: ```suggestion
derive the properties from and override the initially sourced Manifest.

```

    their arguments.

3. `golem.srv.comp.manifest.script.commands : List[String]`

    Specifies a curated list of commands in form of:

    - UTF-8 encoded strings

      No command context or matching mode need to be specified.

      E.g. `["run /bin/cat /etc/motd", "run /bin/date -R"]`

    - UTF-8 encoded JSON strings

      Command context (e.g. `env`) or argument matching mode need to be
      specified for a command.

      E.g. `["{\"run\": { \"args\": \"/bin/date -R\", \"env\": { \"MYVAR\": \"42\", \"match\": \"strict\" }}}"]`

    - mix of both

    `"deploy"`, `"start"` and `"terminate'` commands are always allowed.
    These values become the **default** if no `manifest.script.command` property
    has been set in the Demand, but the `manifest` namespace is present.

4. `golem.srv.comp.manifest.script.match : String`

    Selects a default way of comparing command arguments stated in the manifest
    and the ones received in the ExeScript, unless stated otherwise in a
    command JSON object.

    The `match` property could be one of:

      - `"strict"`: byte-to-byte argument equality (**default**)

>       - `"regex"`: treat arguments as regular expressions (UTF-8 Unicode mode)

Reviewer: Can we please specify the regex standard to be used here, or link to eg. regex syntax spec that will be used?

            "env": {
              "MYVAR": "42"
            },
            "match": "strict"
          }
        }
      ],
      "net.inet.out.protocols": [
        "http",
        "https"
      ],
      "net.inet.out.urls": [
        "http://golemfactory.s3.amazonaws.com/file1",
        "http://golemfactory.s3.amazonaws.com/file2"
      ]
    }
    ```

2. YAML

    ```yaml
    ---
    script:
      match: regex
      commands:
      - run /bin/cat /etc/motd
      - run:
          args: "/bin/date -R"
          env:
            MYVAR: '42'
          match: strict
    net:
      inet:
        out:
          protocols:
          - http
          - https
          urls:
          - http://golemfactory.s3.amazonaws.com/file1
          - http://golemfactory.s3.amazonaws.com/file2
    ```

## Rationale

### External Manifest sourcing

The main purpose of external sourcing is to provide a trusted set of actions
approved by the signing authority (e.g. Golem Factory). The prevalent case of
sourcing would be to download the Manifest included within a Computation
Payload. This way, in most common cases, Requestors won't need to provide a
custom Computation Manifest.

> ### Usefulness of the new property set

Reviewer: ```suggestion

### Usefulness of the new property set
```

---
gap: 4
title: Computation Manifest
description:
author: mf (@mfranciszkiewicz)
status: Draft
type: Feature
---

## Abstract

This GAP introduces a manifest file specification and a set of GCRS (Golem
Computation Resource Standards) properties (herein after referred to as
"Computation Manifest"), which allow Requestors to declare the constraints of
computations they may execute on a Provider's machine.

## Motivation

The need for a Computation Manifest originates from the upcoming Internet
access feature, where each outbound request is routed via Provider's host
operating system. In order to prevent unwanted / illegal behavior on Provider's

> behalf, she needs to remain in control of actions that Requestors can execute

Reviewer 2: ```suggestion
behalf, they need to remain in control of actions that Requestors can execute
```

> on her machine.

Reviewer 2: ```suggestion
on their machine.
```

With Computation Manifests, Requestors constrain themselves to
a certain set of allowed actions, to be negotiated with and approved by a
Provider. Requestors' actions will be verified against the Manifest during
computation.

## Specification

Computation Manifest can take one of the following forms:

- an embedded JSON / YAML object within an external configuration file,
- `golem.srv.comp.manifest` namespace properties, an extension to
  Golem Computation Resource Standards.

The Computation Manifest is expected to be included in a Computation Payload
Manifest ([GAP](https://github.com/golemfactory/golem-architecture/pull/28))
file, where its hash is co-signed with Payload hashes. If a Computation
Manifest is included in the Demand (and thus, Agreement), its role is to
derive the properties from and override the initially sourced Manifest.

> 

Reviewer: Can we please also indicate the "lifecycle" of the Manifest - ie. 
- who creates it?
- what happens with it? (ie. if it is signed and then validated - then by whom, and against what?)
- where does the manifest reside (eg. is it part of the Golem application "package", together with payload GVMIs?)

Author: Done

4. `golem.srv.comp.manifest.script.match : String`

    Selects a default way of comparing command arguments stated in the manifest
    and the ones received in the ExeScript, unless stated otherwise in a
    command JSON object.

    The `match` property could be one of:

      - `"strict"`: byte-to-byte argument equality (**default**)
      - `"regex"`: treat arguments as regular expressions

        Syntax: Perl-compatible regular expressions (UTF-8 Unicode mode),
        w/o the support for look around and backreferences (among others);
        for more information read the documentation of the Rust
        [regex](https://docs.rs/regex/latest/regex/) crate.

5. `golem.srv.comp.manifest.net` (sub-namespace)

    Applies constraints to networking. Currently, outgoing requests to the
    public Internet network are covered.

6. `golem.srv.comp.manifest.net.inet.out.protocols : List[String]`

    List of allowed outbound protocols. Currently **fixed at** `["http", "https"]`.

7. `golem.srv.comp.manifest.net.inet.out.urls : List[String]`

    List of allowed external URLs that outbound requests can be sent to.

>     E.g. `["http://golemfactory.s3.amazonaws.com/file1", "http://golemfactory.s3.amazonaws.com/file2"]`

Reviewer 2: how would we verify a URL on a network-interface level? shouldn't it rather be an IP address? or an IP range?

Author: This question goes well beyond the scope of this GAP :slightly_smiling_face: what you're asking of would be a separate property, not covered here (`net.inet.out.ips` ?).

As of now, filtering is done on OSI L3+ layers (IP and upwards), and domain names have to be resolved (to an address or a range of addresses) before we start executing commands. 

The upsides:
- IP addresses of servers specified in the manifest can change
- developers use human-readable names for destinations

The downsides:
- ExeUnit starts with a delay for DNS name resolution, w/ potential TLS certificate issues
- DNS records can change during execution, so we need a background job to update those (in the future)

    otherwise in a command JSON object.

    The `match` property could be one of:

      - `"strict"`: byte-to-byte argument equality (**default**)
      - `"regex"`: treat arguments as regular expressions

        Syntax: Perl-compatible regular expressions (UTF-8 Unicode mode),
        w/o the support for look around and backreferences (among others);
        for more information read the documentation of the Rust
        [regex](https://docs.rs/regex/latest/regex/) crate.

5. `golem.srv.comp.manifest.net` (sub-namespace)

    Applies constraints to networking. Currently, outgoing requests to the
    public Internet network are covered.

6. `golem.srv.comp.manifest.net.inet.out.protocols : List[String]`

    List of allowed outbound protocols. Currently **fixed at** `["http", "https"]`.

7. `golem.srv.comp.manifest.net.inet.out.urls : List[String]`

    List of allowed external URLs that outbound requests can be sent to.
    E.g. `["http://golemfactory.s3.amazonaws.com/file1", "http://golemfactory.s3.amazonaws.com/file2"]`

#### Example

```json
{
  "golem.srv.comp.manifest.script.match": "regex",
  "golem.srv.comp.manifest.script.commands": [
    "run /bin/cat /etc/motd",

>     "{\"run\": { \"args\": \"/bin/date -R\", \"env\": { \"MYVAR\": \"42\", \"match\": \"strict\" }}}"

Reviewer 3: ```suggestion
    "{\"run\": { \"args\": \"/bin/date -R\", \"env\": { \"MYVAR\": \"42\" }, \"match\": \"strict\" }}"
```

>     "{\"run\": { \"args\": \"/bin/date -R\", \"env\": { \"MYVAR\": \"42\", \"match\": \"strict\" }}}"

Reviewer 3: ```suggestion
    "{\"run\": { \"args\": \"/bin/date -R\", \"env\": { \"MYVAR\": \"42\" }, \"match\": \"strict\" }}"
```
I guess this was the original intention? At least, this version is in `golem.srv.comp.manifest.script.commands` description, and `match` is described as a `way of comparing command arguments`, not env variables.

Although maybe regexp env variables would also be useful?
Or maybe `match` should be a general attribute that might be somehow added to all possible attributes? E.g.
```
"golem.srv.comp.manifest.net.inet.out" = {
    "urls": ["http://golemfactory.s3.amazonaws.com/file*"],
    "match": "regex",
}
```

Reviewer: @mfranciszkiewicz please adjust the GAP content as per the chat on TC call

Author: Done
